### ISSUE TITLE
and filter with exists doesn't work against arrays with duplicate values

### ISSUE THREAD

**Issue by leomchan** [kuzzleio/koncorde#24]
If you have an *and* filter, and one of the sub filters is an *exists* against a array with duplicate values, the other filters are ignored.

## Expected Behavior
The other filters should still be respected.

## Current Behavior
It appears the other filters are ignored.

## Steps to Reproduce
The following code demonstrates the problem

```
const Koncorde = require('koncorde');

const engine = new Koncorde();

const filter = {
  and: [
    {
      equals: {
        name: 'Leo',
      }
    },
    {
      exists: 'skills.languages["javascript"]'
    }
  ]
};

engine.register('index', 'collection', filter)
.then((result) => {
  const matches1 = engine.test('index', 'collection', {
    name: 'Leo',
    skills: {
      languages: ['c', 'java', 'javascript']
    }
  });

  if (matches1.indexOf(result.id) < 0) {
    throw new Error('Document 1 should have matched.');
  }

  const matches2 = engine.test('index', 'collection', {
    name: 'Bob',
    skills: {
      languages: ['pascal', 'javascript', 'python', 'javascript']
    }
  });

  if (matches2.indexOf(result.id) >= 0) {
    throw new Error('Document 2 should not have matched.');
  }
})
.then(() => {
  console.log('Finished.');
  process.exit(0);
})
.catch((err) => {
  console.error(err);
  process.exit(1);
})
```

Note that in the second document `javascript` is duplicated in `skills.languages`. If you remove the extra `javascript` item, the test passes as expected.

## Context (Environment)
I'm using lts/carbon, with koncorde 1.2.1
Problem came across in our internal testing.

**Comment by scottinet** [comment]
Thanks for the heads up. I'll take a look at it.

**Comment by scottinet** [comment]
@leomchan,

Operands (and, or, bool) work by maintaining a counter of **different** conditions that must all be met before a subfilter can match.

That counter should only be decremented once per tested condition on any given subfilter. 
But when testing an array of values with `exists`, the exists matcher decrements the counter without verifying if the condition has already been verified.

For instance, the following condition: `exists: foo["bar"]` matches 2 times if the following object is tested: `{"foo": ["bar", "bar"]}`

This bug is solved by #26, by making the exists matcher test only a set of unique values extracted from the array.

I checked the other matchers and they seem unaffected by this bug.

And... I also activated notifications on this repository. Sorry for not having see that issue earlier.
